File: square.rb
require 'set'

# Represent a single field of the map. These fields are either land or
# unknown - once a square is observed as water, it is deleted
class Square
  @@rows = nil
  @@cols = nil
  @@index = nil

  @@observed = []

  # pretty-print map for debugging
  def self.dump_map(from, to)
    map = ""
    0.upto(@rows - 1) do |row|
      0.upto(@cols - 1) do |col|
        square = Square.at(row, col)

        c = if square.nil?
              'X'
            elsif square == from
              '*'
            elsif square == to
              '$'
            elsif square.observed?
              '_'
            else
              ' '
            end
        map += c
      end

      map += "\n"
    end

    map
  end

  # Build the full grid, assuming no water. As we discover water, squares will be deleted
  def self.create_squares(rows, cols, viewradius2)
    @@viewradius2 = viewradius2
    @@rows = rows
    @@cols = cols

    @@visibility_mask = []
    viewradius = Math.sqrt(@@viewradius2).ceil
    (-1 * viewradius).upto(viewradius).each do |row_offset|
      (-1 * viewradius).upto(viewradius).each do |col_offset|
        if distance2(0, 0, row_offset, col_offset) < @@viewradius2
          @@visibility_mask << [row_offset, col_offset]
        end
      end
    end

    @@index = Array.new(@@rows * @@cols) do |i|
      Square.new(i / @@cols, i % @@cols)
    end
  end

  def self.all
    @@index.compact
  end

  def self.observed
    @@observed
  end

  def self.reset!
    @@index.each { |square| square.reset! if square }
  end

  def self.at(row, col)
    @@index[position_to_index(row, col)]
  end

  def self.position_to_index(row, col)
    ((row % @@rows) * @@cols) + (col % @@cols)
  end

  attr_reader :row, :col
  attr_accessor :ant, :next_ant
  attr_accessor :item
  attr_reader :goals

  def initialize(row, col)
    @row = row
    @col = col
    @observed = false
    @visited = false
    @item = nil

    # Map of a goal instance to a route to that goal, from this square
    @goals = {} # TODO find a clever way to work Wander into here
  end

  def neighbors
    @neighbors ||= make_neighbors
    Set.new(@neighbors.values)
  end

  def direction_to_neighbors
    @neighbors ||= make_neighbors
  end

  def make_neighbors
    possible_neighbors = [[:e, 0, 1], [:w, 0, -1], [:s, 1, 0], [:n, -1, 0]].map do |direction, row_offset, col_offset|
      neighbor_row = @row + row_offset
      neighbor_col = @col + col_offset
      [direction, Square.at(neighbor_row, neighbor_col)]
    end

    actual_neighbors = possible_neighbors.reject { |direction, neighbor| neighbor.nil? }

    Hash[actual_neighbors]
  end

  def direction_to(destination)
    direction_to_neighbors.find { |direction, neighbor| neighbor == destination }.first
  end

  # Mark as visited and return the number of observed squares
  def visit!
    return 0 if @visited
    @visited = true

    visible_squares.inject(0) do |observed_count, vis|
      if vis.observed?
        observed_count # no change
      else
        vis.observe!
        observed_count + 1
      end
    end
  end

  def visited?
    @visited
  end

  def observed?
    @observed
  end

  def observe!
    @observed = true
    @@observed << self
  end

  def visible_squares
    @_visible_squares ||= @@visibility_mask.map do |row_offset, col_offset|
      Square.at(@row + row_offset, @col + col_offset)
    end.compact

    # Can't cache which ones are non-nil
    @_visible_squares.find_all { |s| Square.at(s.row, s.col) }
  end

  def visible(square)
    distance2(square) < @@viewradius2
  end

  def frontier?
    observed? && ! neighbors.all?(&:observed?)
  end

  def has_food?
    @item && @item.kind_of?(Food)
  end

  def has_hill?
    @item && @item.kind_of?(Hill)
  end

  def has_enemy_ant?
    @item && @item.kind_of?(EnemyAnt)
  end

  def destroy!
    neighbors.each do |neighbor|
      neighbor.remove_dead_neighbor(self)
    end

    @@index[Square.position_to_index(@row, @col)] = nil
    @@observed.delete(self)
  end

  def remove_dead_neighbor(dead_neighbor)
    direction = direction_to(dead_neighbor)
    @neighbors.delete(direction)
  end

  def self.distance2(r1, c1, r2, c2)
    rdelt = (r1 - r2).abs
    cdelt = (c1 - c2).abs
    dr = [rdelt, @@rows - rdelt].min
    dc = [cdelt, @@cols - cdelt].min
    (dr**2 + dc**2)
  end

  def distance2(other)
    Square.distance2(@row, @col, other.row, other.col)
  end

  def to_s
    "[#{@row}, #{@col}]"
  end

  def blacklist
    Set.new(neighbors.find_all { |n| n.next_ant || n.has_food? || n.has_hill? })
  end

  def reset!
    @next_ant = @ant = nil
  end
end

File: java/src/com/bradleybuda/Square.java
package com.bradleybuda;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

import org.apache.commons.lang.builder.EqualsBuilder;
import org.apache.commons.lang.builder.HashCodeBuilder;

public class Square {
    private enum Direction {
        EAST('e'), WEST('w'), SOUTH('s'), NORTH('n');
        
        private final char symbol;
        private Direction(final char symbol) {
            this.symbol = symbol;
        }
        
        private Offset offset() {
            switch (this) {
            case EAST:
                return new Offset(0, 1);
            case WEST:
                return new Offset(0, -1);
            case SOUTH:
                return new Offset(1, 0);
            case NORTH:
                return new Offset(-1, 0);
            } 
        }
    }
    
    private static class Offset {
        int row;
        int col;

        Offset(int row, int col) {
            this.row = row;
            this.col = col;
        }
    }
    
    private static int ROWS;
    private static int COLS;
    private static int VIEW_RADIUS_2;
    private static Square[][] INDEX;
    public static final Set<Square> OBSERVED = new HashSet<Square>();
    private static final Set<Offset> VISIBLITY_MASK = new HashSet<Offset>();
    
    public static String dumpMap(final Square from, final Square to) {
        final StringBuilder map = new StringBuilder();
        
        for (int row = 0; row < ROWS; ++row) {
            for (int col = 0; col < COLS; ++col) {
                final Square square = INDEX[row][col];
                
                char symbol;
                if (square == null)
                    symbol = 'X';
                else if (square == from)
                    symbol = '*';
                else if (square == to)
                    symbol = '$';
                else if (square.isObserved())
                    symbol = '_';
                else
                    symbol = ' ';
                map.append(symbol);
            }
        }
        
        return map.toString();
    }
    
    public static void createSquares(int rows, int cols, int viewRadius2) {
        Square.VIEW_RADIUS_2 = viewRadius2;
        Square.ROWS = rows;
        Square.COLS = cols;
     
        int viewRadius = (int)Math.ceil(Math.sqrt(VIEW_RADIUS_2));
        for (int rowOffset = -1 * viewRadius; rowOffset <= viewRadius; ++rowOffset)
            for (int colOffset = -1 * viewRadius; rowOffset <= viewRadius; ++rowOffset)
                if (Square.distance2(0, 0, rowOffset, colOffset) < VIEW_RADIUS_2)
                    VISIBLITY_MASK.add(new Offset(rowOffset, colOffset));
        
        INDEX = new Square[ROWS][COLS];
        for (int row = 0; row < ROWS; ++row)
            for (int col = 0; col < COLS; ++col)
                INDEX[row][col] = new Square(row, col);
    }
    
    public static Set<Square> all() {
        final Set<Square> all = new HashSet<Square>();
        for (int row = 0; row < ROWS; ++row)
            for (int col = 0; col < COLS; ++col)
                if (INDEX[row][col] != null)
                    all.add(INDEX[row][col]);
        
        return all;
    }
    
    public static void resetAll() {
        for (int row = 0; row < ROWS; ++row)
            for (int col = 0; col < COLS; ++col)
                if (INDEX[row][col] != null)
                    INDEX[row][col].reset();        
    }
    
    public static Square at(int row, int col) {
        return INDEX[normalizeRow(row)][normalizeCol(col)];
    }
    
    private static int normalizeRow(int row) {
        return row % ROWS;
    }
    
    private static int normalizeCol(int col) {
        return col % COLS;
    }
    
    private final int row;
    private final int col;
    private boolean observed = false;
    private boolean visited = false;
    private Item item = null;
    private final Map<Goal, Route> goals = new HashMap<Goal, Route>();
    private final Map<Direction, Square> neighbors = new HashMap<Direction, Square>(4);
    private Set<Square> visibleSquares;
    private Ant ant;
    private Ant nextAnt;    
    
    public Square(int row, int col) {
        this.row = row;
        this.col = col;
    }
    
    public Set<Square> neighbors() {
        if (neighbors == null)
            makeNeighbors();
        return new HashSet<Square>(neighbors.values());
    }
    
    public Map<Direction, Square> directionToNeighbors() {
        if (neighbors == null)
            makeNeighbors();
        return neighbors;
    }
    
    private void makeNeighbors() {
        for (final Direction d : Direction.values()) {
            final Offset o = d.offset();
            final int neighborRow = row + o.row;
            final int neighborCol = col + o.col;
            final Square neighbor = Square.at(neighborRow, neighborCol);
            if (neighbor != null)
            neighbors.put(d, neighbor);
        }
    }
    
    public Direction directionTo(final Square neighbor) {
        for (final Map.Entry<Direction, Square> entry : neighbors.entrySet()) {
            if (entry.getValue().equals(neighbor)) {
                return entry.getKey();
            }
        }
    }
    
    public boolean isVisited() {
        return visited;
    }
    
    public boolean isObserved() {
        return observed;
    }
    
    public void observe() {
        observed = true;
        OBSERVED.add(this);
    }
    
    public Set<Square> visibleSquares() {
        // initial memoized list
        if (visibleSquares == null) {
            visibleSquares = new HashSet<Square>(VISIBLITY_MASK.size());
            for (final Offset o : VISIBLITY_MASK) {
                final Square s = Square.at(row + o.row, col + o.col);
                if (s != null) {
                    visibleSquares.add(s);
                }
            }
        }
        
        // double-check memoized against ones that have disappeared
        final Set<Square> retVal = new HashSet<Square>(visibleSquares.size());
        for (final Square s : visibleSquares) {
            final Square stillThere = Square.at(s.row, s.col);
            if (stillThere != null)
                retVal.add(stillThere);
        }
        
        return retVal;
    }
    
    public boolean visible(Square square) {
        return distance2(square) < VIEW_RADIUS_2; 
    }
    
    public boolean isFrontier() {
        if (!isObserved()) {
            return false;
        } else {
          for (Square s : neighbors())
              if (!s.isObserved())
                  return true; // observed but has unobserved neighbor
          return false; // observed and all neighbors observed
        }
    }
    
    public boolean hasFood() {
        return (item != null) && (item instanceof Food);
    }
    
    public boolean hasHill() {
        return (item != null) && (item instanceof Hill);
    }

    public boolean hasEnemyAnt() {
        return (item != null) && (item instanceof EnemyAnt);
    }
    
    public void destroy() {
        for (final Square neighbor : neighbors())
            neighbor.removeDeadNeighbor(this);
        INDEX[row][col] = null;
        OBSERVED.remove(this);
    }
    
    private void removeDeadNeighbor(Square deadNeighbor) {
        final Direction direction = directionTo(deadNeighbor);
        neighbors.remove(direction);
    }
    
    public static int distance2(int r1, int c1, int r2, int c2) {
        final int rdelt = Math.abs(r1 - r2);
        final int cdelt = Math.abs(c1 - c2);
        final int dr = Math.min(rdelt, ROWS - rdelt);
        final int dc = Math.min(cdelt, COLS - cdelt);
        return (dr ^ 2) + (dc ^ 2);
    }
    
    public int distance2(Square other) {
        return Square.distance2(row, col, other.row, other.col);
    }
    
    @Override
    public String toString() {
        return String.format("[%d, %d]", row, col);
    }
    
    public Set<Square> blacklist() {
        final Set<Square> blacklist = new HashSet<Square>();
        for (final Square neighbor : neighbors()) {
            if (neighbor.nextAnt != null)
                blacklist.add(neighbor);
            else if (neighbor.hasFood())
                blacklist.add(neighbor);
            else if (neighbor.hasHill() && neighbor.item.isMine())
                blacklist.add(neighbor);
        }
    }
    
    public void reset() {
        nextAnt = ant = null;
    }
    
    @Override
    public int hashCode() {
        return new HashCodeBuilder().append(row).append(col).hashCode();
    }
    
    @Override
    public boolean equals(Object obj) {
        final Square other = (Square)obj;
        return new EqualsBuilder().append(row, other.row).append(col, other.col).isEquals();
    }
}
